linkerd/app/core/src/serve.rs

```diff
@@ -15,36 +16,47 @@ pub trait HasSpan {
 /// connection-accepting service.
 ///
 /// The task is driven until the provided `drain` is notified.
-pub async fn serve<L, A>(listen: L, accept: A, drain: drain::Watch) -> Result<(), Error>
+pub async fn serve<A, C>(
+    listen: impl Stream<Item = std::io::Result<C>>,
+    accept: A,
+) -> Result<(), Error>
 where
-    L: Listen + Send + 'static,
-    L::Connection: HasSpan,
-    L::Error: std::error::Error + Send + 'static,
-    A: Accept<L::Connection> + Send + 'static,
-    A::Error: Send + 'static,
+    C: HasSpan,
+    A: Accept<C> + Send + 'static,
     A::Future: Send + 'static,
-    A::Error: Into<Error>,
+    A::Error: Into<Error> + Send + 'static,
     A::ConnectionFuture: Send + 'static,
+    A::ConnectionError: Send + 'static,
 {
-    // As soon as we get a shutdown signal, the listener task completes and
-    // stops accepting new connections.
-    debug!(listen.addr = %listen.listen_addr(), "serving");
-
     // Initialize tracing & log errors on the accept stack.
-    let accept = TraceAccept {
+    let mut accept = TraceAccept {
         accept: AcceptError::new(accept),
         span: Span::current(),
-    };
+    }
+    .into_service();
 
-    // Drive connections from the listener into the accept stack until a drain
-    // is signaled.
-    tokio::select! {
-        err = listen.serve(accept) => {
-            // The listener may fail but it may not complete otherwise.
-            match err? {}
+    futures::pin_mut!(listen);
+    loop {
+        let ready = accept.ready_and().await.map_err(Into::<Error>::into)?;
+        match listen.next().await {
+            None => return Ok(()),
+            Some(conn) => {
```

Comment:
I may be wrong, but is it possible that driving the accept service to readiness _before_ waiting for the next incoming connection could be problematic, along the same lines as how the buffer worker was causing load balancers to miss discovery updates?

Should we do something more like 
https://github.com/linkerd/linkerd2-proxy/blob/1b3e899efad2ad4ee471c5b49482437e7a325bde/linkerd/buffer/src/dispatch.rs#L47-L51
instead, or is that not an issue here?

Reply:
You're probably right that it's safer to drive it once the connection is received.

Reply:
Yeah, I didn't see anything that looked obviously broken as a result of this, but it's bitten us before...